### PR TITLE
Cms

### PR DESCRIPTION
**Pull request by bethios**
screenshot 2017-06-11 13 30 44

https://sacred-art-development.herokuapp.com/

### CODE REVIEW

**Comment by bethios** [comment]
[(M) As an admin I want to be able to log into sacred art's site so that I can access the CMS](https://trello.com/c/VcQwhLDl/69-m-as-an-admin-i-want-to-be-able-to-log-into-sacred-arts-site-so-that-i-can-access-the-cms)

**Comment by bethios** [comment]
[(S) As a visitor on the artists page I want to see the category listed at the top of the page so that I know which category I'm viewing.](https://trello.com/c/qKnPmZJF/39-s-as-a-visitor-on-the-artists-page-i-want-to-see-the-category-listed-at-the-top-of-the-page-so-that-i-know-which-category-im-vie)

**Comment by bethios** [comment]
[(M) As a visitor on the categories page I want to click on my chosen category so that I can be taken to a list of artists.](https://trello.com/c/YZFBivMz/35-m-as-a-visitor-on-the-categories-page-i-want-to-click-on-my-chosen-category-so-that-i-can-be-taken-to-a-list-of-artists)

**Comment by bethios** [comment]
[(M) As an admin on the artists page I want to delete artists so that I can update the site.](https://trello.com/c/3zjEQ427/38-m-as-an-admin-on-the-artists-page-i-want-to-delete-artists-so-that-i-can-update-the-site)

**Comment by bethios** [comment]
[(M) As an admin on the artists page I want to edit artists so that I can update the site.](https://trello.com/c/0He2K6VJ/65-m-as-an-admin-on-the-artists-page-i-want-to-edit-artists-so-that-i-can-update-the-site)

**Comment by bethios** [comment]
[(M) As an admin on the artists page I want to add new artists so that I can include them on the site](https://trello.com/c/tX0vFOcy/37-m-as-an-admin-on-the-artists-page-i-want-to-add-new-artists-so-that-i-can-include-them-on-the-site)

**Comment by bethios** [comment]
[(M) As a visitor on the artists page I want to read information and see photos of each artist so that I can learn more about them.](https://trello.com/c/hFBxghrp/36-m-as-a-visitor-on-the-artists-page-i-want-to-read-information-and-see-photos-of-each-artist-so-that-i-can-learn-more-about-them)

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -170,8 +171,11 @@ iframe{
   margin: 8px;
   font-size: 1.1em;
   background-color: $slate;
-  color: white;
+  color: $white;
+}
 
+.footer p{
+  font-size: 1em;
```
Comment (cblant20) [inline app/assets/stylesheets/application.scss:178]: noooo font-size needs to be a variable

File: app/assets/stylesheets/artists.scss [hunk 2]
```diff
@@ -1,3 +1,89 @@
...
+
+  .artist{
+    grid-template-columns: repeat(11, 1fr);
+    grid-template-areas:
+            "img1 img1 img1 img1 img2 img2 nm  nm  nm  nm  nm"
+            "img1 img1 img1 img1 img2 img2 bio bio bio bio bio"
+            "img1 img1 img1 img1 img3 img3 bio bio bio bio bio"
+            "img1 img1 img1 img1 img3 img3 bio bio bio bio bio";
+  }
+}
```
Comment (cblant20) [inline app/assets/stylesheets/artists.scss:89]: newline

File: app/assets/stylesheets/welcome.scss [hunk 3]
```diff
@@ -114,15 +109,13 @@ iframe{
...
   }
 }
 
-
-
+@media screen and (min-width: 1200px){
+  .instagram-feed{
+    grid-template-columns: 25% 25% 25% 25%;
+  }
+}
```
Comment (cblant20) [inline app/assets/stylesheets/welcome.scss:121]: newline

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -2,4 +2,14 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+  include SessionsHelper
+
+  private
+
+  def require_sign_in
+    unless current_user
+      flash[:alert] = "You must be logged in to do that"
```
Comment (cblant20) [inline app/controllers/application_controller.rb:11]: good

File: app/controllers/artists_controller.rb [hunk 5]
```diff
@@ -1,23 +1,76 @@
...
+    @artist.body = params[:artist][:body]
+    @artist.main_image = params[:artist][:main_image]
+    @artist.image_2 = params[:artist][:image_2]
+    @artist.image_3 = params[:artist][:image_3]
+
+    if @artist.save
+      flash[:notice] = "Artist was updated."
+      redirect_to [@artist.category]
+    else
+      flash.now[:alert] = "There was an error saving the artist. Please try again."
```
Comment (cblant20) [inline app/controllers/artists_controller.rb:43]: good

File: app/views/artists/_each.html.erb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+          <%= image_tag(artist.image_2.url(:small), :class => 'artistSecondImage') %>
+      <% end %>
+      <% if artist.image_3 %>
+          <%= image_tag(artist.image_3.url(:small), :class => 'artistThirdImage') %>
+      <% end %>
+
+      <p class="artistName"><strong><%= artist.name %></strong></p>
+      <p class="bioBody"><%= artist.body %></p>
+    </div>
+<% end %>
```
Comment (cblant20) [inline app/views/artists/_each.html.erb:22]: newline

File: app/views/artists/_each.html.erb [hunk 7]
```diff
@@ -0,0 +1,22 @@
+<% @category.artists.each do |artist| %>
+    <div class="content-container wrapper artist">
+      <% if current_user && current_user.role == 'admin' %>
+          <%= link_to "Edit Artist", edit_category_artist_path(artist.category, artist), class: 'btn btn-default' %>
+          <%= link_to "Delete Artist", [artist.category, artist], method: :delete, class: 'btn btn-danger',
+                      data: { confirm: 'Are you sure you want to delete this recipe?' } %>
```
Comment (cblant20) [inline app/views/artists/_each.html.erb:6]: 'artist'

File: app/views/artists/_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    <%= f.file_field :image_2 %>
+  </div>
+  <div class="form-group">
+    <%= f.label :image_3 %>
+    <%= f.file_field :image_3 %>
+  </div>
+  <div class="form-group">
+    <%= f.submit "Save", class: 'btn btn-success' %>
+  </div>
+<% end %>
```
Comment (cblant20) [inline app/views/artists/_form.html.erb:25]: newline

File: app/views/artists/_form.html.erb [hunk 9]
```diff
@@ -0,0 +1,25 @@
+<%= form_for [category, artist], html: { multipart: true } do |f| %>
```
Comment (cblant20) [inline app/views/artists/_form.html.erb:1]: responsivify?

File: app/views/artists/edit.html.erb [hunk 10]
```diff
@@ -1,2 +1,4 @@
-<h1>Artists#edit</h1>
-<p>Find me in app/views/artists/edit.html.erb</p>
+<h2>Edit Artist Info</h2>
+
+<%= render partial: 'artists/form', locals: {category: @artist.category, artist: @artist} %>
```
Comment (cblant20) [inline app/views/artists/edit.html.erb:3]: the images don't prepopulate

File: app/views/layouts/application.html.erb [hunk 11]
```diff
@@ -43,6 +43,21 @@
     <li class="nav-li"><%= link_to "CONTACT", contact_path %></li>
   </ul>
 </nav>
+<% if current_user %>
+    <%= current_user.name %> - <%= link_to "Sign Out", session_path(current_user), method: :delete %>
+<% end %>
+
+<% if flash[:notice] %>
```
Comment (cblant20) [inline app/views/layouts/application.html.erb:50]: ?

File: app/views/sessions/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+        <div class="form-group">
+          <%= f.label :password %>
+          <%= f.password_field :password, class: 'form-control', placeholder: "Enter password" %>
+        </div>
+        <div class="form-group">
+          <%= f.submit "Sign in", class: 'btn btn-success' %>
+        </div>
+    <% end %>
+  </div>
+</div>
```
Comment (cblant20) [inline app/views/sessions/new.html.erb:19]: newline

File: app/views/users/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+        <div class="form-group">
+            <%= f.label :password_confirmation %>
+            <%= f.password_field :password_confirmation, class: 'form-control', placeholder: "Enter password confirmation" %>
+        </div>
+        <div class="form-group">
+          <%= f.submit "Sign up", class: 'btn btn-success' %>
+        </div>
+    <% end %>
+  </div>
+</div>
```
Comment (cblant20) [inline app/views/users/new.html.erb:35]: newline

File: app/views/users/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,35 @@
+<div class="row">
```
Comment (cblant20) [inline app/views/users/new.html.erb:1]: you're allowing new users?

File: spec/controllers/artists_controller_spec.rb [hunk 15]
```diff
@@ -1,53 +1,150 @@
...
 RSpec.describe ArtistsController, type: :controller do
+  let(:my_category) { Category.create!( name: "Original Art", image: "original.jpg") }
+  let(:my_artist) { my_category.artists.create!(name: 'test artist', body: 'test artist body', category_id: my_category.id) }
+  let(:admin) { User.create!(name: 'admin user', email: 'admin@admin.com', password: "helloworld", role: :admin) }
 
-  describe "GET #new" do
-    it "returns http success" do
-      get :new
-      expect(response).to have_http_status(:success)
+  context "guest" do
```
Comment (cblant20) [inline spec/controllers/artists_controller_spec.rb:9]: I think you should use a 'when' here

File: spec/controllers/artists_controller_spec.rb [hunk 16]
```diff
@@ -1,53 +1,150 @@
...
+        count = Artist.where({id: my_artist.id}).size
+        expect(count).to eq 0
+      end
+
+      it "redirects to artist index" do
+        delete :destroy, category_id: my_category.id, id: my_artist.id
+        expect(response).to redirect_to my_category
+      end
     end
   end
```
Comment (cblant20) [inline spec/controllers/artists_controller_spec.rb:149]: do you time out the session ever?

File: spec/controllers/sessions_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+      expect(session[:user_id]).to be_nil
+    end
+
+    it "flashes #error with bad email address" do
+      post :create, session: {email: "does not exist"}
+      expect(flash.now[:alert]).to be_present
+    end
+
+    it "renders #new with bad email address" do
+      post :create, session: {email: "does not exist"}
```
Comment (cblant20) [inline spec/controllers/sessions_controller_spec.rb:35]: what about a bad pass?

File: spec/models/artist_spec.rb [hunk 18]
```diff
@@ -1,5 +1,20 @@
...
+  let(:body){ 'test artist body' }
+  let(:category) { Category.create!( name: "Original Art", image: "original.jpg") }
+  let(:category_id) { category.id}
+
+  let(:artist) { category.artists.create!(name: name, body: body, category_id: category_id) }
+
+  it { is_expected.to belong_to(:category)}
+
+  describe "attributes" do
+    it "has name and body attributes" do
```
Comment (cblant20) [inline spec/models/artist_spec.rb:16]: what if it doesn't?

File: app/assets/stylesheets/artists.scss (hunk 2 above)
Comment (cblant20) [reply app/assets/stylesheets/artists.scss:89]: Forgot I have to click submit review! Sorry! Did these yesterday and just now remembered!

**Pull request merged**